Projects/b_u585i_iot02a_ntz/Src/ota_pal/ota_pal_stm32u5_ntz.c

```diff
@@ -47,150 +47,306 @@
+    .xObLaunchPending = pdFALSE,
+};
+
+static uint32_t ulBankAtBootup = 0;
+
+/* Static function forward declarations */
 
-    lfs_file_t xFile = { 0 };
+/* Load/Save/Delete */
+static BaseType_t prvInitializePalContext( OtaPalContext_t * pxContext );
+static BaseType_t prvWritePalNvContext( OtaPalContext_t * pxContext );
+static BaseType_t prvDeletePalNvContext( void );
+static OtaPalContext_t * prvGetImageContext( void );
 
-    /* Open the file */
-    lReturn = lfs_file_open( pLfsCtx, &xFile, pcFileName, LFS_O_RDONLY );
+/* Active / Inactive bank helpers */
+static uint32_t prvGetActiveBank( void );
+static uint32_t prvGetInactiveBank( void );
 
-    /* Read the header */
-    if( lReturn == LFS_ERR_OK )
+/* STM32 HAL Wrappers */
+static HAL_StatusTypeDef prvFlashSetDualBankMode( void );
+static BaseType_t prvSelectBank( uint32_t ulNewBank );
+static uint32_t prvGetBankSettingFromOB( void );
+
+static void prvOptionByteApply( void );
+
+/* Flash write./erase */
+static HAL_StatusTypeDef prvWriteToFlash( uint32_t destination,
+                                          uint8_t * pSource,
+                                          uint32_t length );
+
+static BaseType_t prvEraseBank( uint32_t bankNumber );
+
+/* Verify signature */
+static OtaPalStatus_t prvValidateSignature( const char * pcPubKeyLabel,
+                                            const unsigned char * pucSignature,
+                                            const size_t uxSignatureLength,
+                                            const unsigned char * pucImageHash,
+                                            const size_t uxHashLength );
+
+static BaseType_t xCalculateImageHash( const unsigned char * pucImageAddress,
+                                       const size_t uxImageLength,
+                                       unsigned char * pucHashBuffer,
+                                       size_t uxHashBufferLength,
+                                       size_t * puxHashLength );
+
+static const char * pcPalStateToString( OtaPalState_t xPalState )
+{
+    const char * pcStateString = "";
+
+    if( xPalState <= OTA_PAL_INVALID )
     {
-        lReturn = lfs_file_read( pLfsCtx, &xFile, pContext, sizeof( OtaImageContext_t ) );
+        pcStateString = ppcPalStateString[ xPalState ];
+    }
 
-        if( lReturn == sizeof( OtaImageContext_t ) )
-        {
-            status = pdTRUE;
-        }
-        else
-        {
-            LogError( ( " Failed to read OTA image context from file %s, error = %d.\r\n ", pcFileName, lReturn ) );
-        }
+    return pcStateString;
+}
 
-        ( void ) lfs_file_close( pLfsCtx, &xFile );
+static inline uint32_t ulGetOtherBank( uint32_t ulBankNumber )
+{
+    uint32_t ulRevertBank;
+
+    if( ulBankNumber == FLASH_BANK_2 )
+    {
+        ulRevertBank = FLASH_BANK_1;
+    }
+    else if( ulBankNumber == FLASH_BANK_1 )
+    {
+        ulRevertBank = FLASH_BANK_2;
+    }
+    else
+    {
+        ulRevertBank = 0;
```

Comment:
Should probably assert on this condition so as to not conflict with definition of `FLASH_BANK_1` and `FLASH_BANK_2` in future.